### ISSUE TITLE
Each run of `test.sh` is very slow

### ISSUE THREAD

**Issue by carlosgmartin** [google-deepmind/optax#1091]
On my machine, each run of `sh test.sh` takes about 4.5 minutes, which is inconvenient for development.

One potential way to make it faster could to keep a gitignored `test_venv` directory at the root of the repo that persists from run to run (that is, is _not_ cleaned up by `test.sh`), so that all the `pip install` packages used by `test.sh` don't need to be reinstalled on each run.

Doing this brings the runtime down to 1 minute on my machine. I can submit a PR demonstrating the changes, if desired. I'm also happy to hear any other suggestions.

**Comment by vroulet** [comment]
I am surprised that the issue could be the installation of packages.
To me what slows down the tests a lot is (i) pytype, (ii) doc building and doctests.
However, by keeping the test_venv, some of theses processes (like ptype) may keep a cache that accelerates subsequent tests.

In any case, if you have a solution that brings the tests down to 1 minute locally, we may at least add an option like a --local flag to prevent the cleanup and add the test_venv to the gitignored files.

@fabianp may also know better on such issues.

**Comment by carlosgmartin** [comment]
@vroulet Yes, that kind of caching might be contributing to the speedup as well.

**Comment by fabianp** [comment]
we could also have different scripts for {pytype, tests, building and testing documentation}, this way they could be run in parallel in github actions

**Comment by carlosgmartin** [comment]
@fabianp @vroulet I've suggested some related changes at #1110.

**Comment by rdyro** [comment]
This should be addressed now.
